code/part-one/tests/04-Mining.js

```diff
@@ -0,0 +1,140 @@
+      recipient = signing.getPublicKey(signing.createPrivateKey());
+      amount = Math.ceil(Math.random() * 100);
+      transaction = new MineableTransaction(signer, recipient, amount);
+    });
+
+    it('should include signer public key as source', function() {
+      expect(transaction.source).to.equal(signing.getPublicKey(signer));
+    });
+
+    it('should include the passed recipient and amount', function() {
+      expect(transaction.recipient).to.equal(recipient);
+      expect(transaction.amount).to.equal(amount);
+    });
+
+    it('should include a valid signature', function() {
+      const { source, signature } = transaction;
+      const signedMessage = source + recipient + amount;
+
+      expect(signing.verify(source, signedMessage, signature)).to.be.true;
+    });
+
+    it('should create a reward by making signer the recipient', function() {
+      const reward = new MineableTransaction(signer, null, amount);
+
+      expect(reward.source).to.equal(null);
+      expect(reward.recipient).to.equal(signing.getPublicKey(signer));
+    });
+  });
+
+  describe('MineableBlock', function() {
+    let previousHash = null;
+    let transactions = null;
+    let block = null;
+
+    beforeEach(function() {
+      const signer = signing.createPrivateKey();
+      const recipient = signing.getPublicKey(signing.createPrivateKey());
+      const amount = Math.ceil(Math.random() * 100);
+
+      transactions = [ new MineableTransaction(signer, recipient, amount) ];
+      previousHash = randomBytes(64).toString('hex');
+
+      block = new MineableBlock(transactions, previousHash);
+    });
+
+    it('should be instantiated without a hash', function() {
+      expect(!!block.hash).to.be.false;
+    });
+
+  });
+
+  describe('MineableChain', function() {
+    let blockchain = null;
+    let transaction = null;
+    let miner = null;
+
+    beforeEach(function() {
+      blockchain = new MineableChain();
+
+      const signer = signing.createPrivateKey();
+      const recipient = signing.getPublicKey(signing.createPrivateKey());
+      const amount = Math.ceil(Math.random() * 100);
+      transaction = new MineableTransaction(signer, recipient, amount);
+
+      miner = signing.createPrivateKey();
+      blockchain.addTransaction(transaction);
+      blockchain.mine(miner);
+    })
+
+    it('should have properties for difficulty and reward', function() {
+      expect(blockchain.difficulty).to.exist.and.be.a('number');
+      expect(blockchain.reward).to.exist.and.be.a('number');
+    });
+
+    it('should be able to mine new blocks', function() {
+      expect(blockchain.blocks).to.have.lengthOf(2);
+
+      const head = blockchain.getHeadBlock();
+      expect(head.transactions).to.deep.include(transaction);
+    });
+
+    it('should mine blocks with a valid zero-leading hash', function() {
+      const head = blockchain.getHeadBlock();
+      const zeros = Array(blockchain.difficulty.length).fill('0').join('');
```

Comment:
1. I think `blockchain.difficulty.length` should actually be `blockchain.difficulty`.
2. We can try `'0'.repeat(blockchain.difficulty)` here too